proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,217 @@
+* Proposal: [0010](0010-resource-element-type-validation.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: Joshua Batista
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
+
+## Introduction
+Resources are often used in HLSL, with various resource element types.
+
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+In this code, the element type is `float`, and the resource type is `RWBuffer`.
+`RWBuffer`, along with some other buffers and textures, fall under the "typed buffer"
+category. Other buffer resources are categorized as "raw buffers". 
+Below is a description of resources that are considered "typed buffers" vs "raw buffers".
+* typed buffers
+  * [RW|RasterizerOrdered]Buffer
+  * [Feedback]Texture*
+* raw buffers
+  * [Append|Consume|RW|RasterizerOrdered]StructuredBuffer
+  * [RW]ByteAddressBuffer
+
+There is a distinct set of rules that define valid element types for typed buffer resources
+and valid element types for raw buffer resources.
+
+Element types for typed buffer resources:
+* Are not intangible (e.g., isn't a resource type)
+* Must be vectors or scalars of arithmetic types, not bools nor enums nor arrays
+* Should be a scalar or homogenous vector of a floating-point or integer type, with a maximum of 4 components after translating 64-bit components into pairs of uint32_t components
+Element types for raw buffer resources:
+* Are not intangible (e.g., isn't a resource type)
+* All constituent types must be arithmetic types or bools or enums
+
+Resource types are never allowed as element types (i.e., `RWBuffer<int>` as an element type).
+If someone writes `RWBuffer<MyCustomType>` and MyCustomType is not a valid element type, 
+there should be infrastructure to reject this element type and emit a message explaining 
+why it was rejected as an element type.

```

Comment:
Isn't the intangible property sufficient, plus saying that user defined types are made intangible when containing intangible types.

Reply:
This is indeed repeating the same information, but I think it's a simpler, alternative way to convey more information, that draws the reader to a conclusion they may not have come to.

Reply:
The point of this is to provide a roadmap for implementation, not documentation for users. Simplifying the language and not repeating things should be the goal.

We should be able to say that cannot be intangible types. We don't even need to go into details about what an intangible type is, we can just link to the language spec for that information.